### PR TITLE
Use 8x8 VRS texel size where supported

### PR DESCRIPTION
**Pull request by AlbinBernhardssonARM**
Sample texel center coordinates when generating SRI.

When the VRS extension is present but either pipeline or attachment VRS is not supported, disable VRS instead of failing device creation.

Use 8x8 VRS tiles where supported. When using 8x8, each thread during SRI generation handles a 2x2 region, instead of 2x4. This is to increase the number of threads per workgroup, to get better utilization.

### CODE REVIEW

File: AnKi/Gr/Vulkan/GrManagerImpl.cpp [hunk 1]
```diff
@@ -1012,16 +1012,43 @@ Error GrManagerImpl::initDevice(const GrManagerInitInfo& init)
...
-		ci.pNext = &m_fragmentShadingRateFeatures;
+		VkPhysicalDeviceFragmentShadingRatePropertiesKHR fragmentShadingRateProperties = {};
+		fragmentShadingRateProperties.sType = VK_STRUCTURE_TYPE_PHYSICAL_DEVICE_FRAGMENT_SHADING_RATE_PROPERTIES_KHR;
+
+		VkPhysicalDeviceProperties2 properties = {};
+		properties.sType = VK_STRUCTURE_TYPE_PHYSICAL_DEVICE_PROPERTIES_2;
+		properties.pNext = &fragmentShadingRateProperties;
+		vkGetPhysicalDeviceProperties2(m_physicalDevice, &properties);
+
+		if(fragmentShadingRateProperties.minFragmentShadingRateAttachmentTexelSize.width > 16
```
Comment (godlikepanos) [inline AnKi/Gr/Vulkan/GrManagerImpl.cpp:1031]: I guess that will disable VRS on certain platforms that go up to 32x32. I guess this check should be if it doesn't support 16x16 or it doesn't support 8x8.

File: AnKi/Gr/Vulkan/GrManagerImpl.cpp [hunk 2]
```diff
@@ -1012,16 +1012,43 @@ Error GrManagerImpl::initDevice(const GrManagerInitInfo& init)
...
-			return Error::FUNCTION_FAILED;
+			m_capabilities.m_vrs = false;
 		}
 
 		// Disable some things
 		m_fragmentShadingRateFeatures.primitiveFragmentShadingRate = false;
 
-		m_fragmentShadingRateFeatures.pNext = const_cast<void*>(ci.pNext);
-		ci.pNext = &m_fragmentShadingRateFeatures;
+		VkPhysicalDeviceFragmentShadingRatePropertiesKHR fragmentShadingRateProperties = {};
```
Comment (godlikepanos) [inline AnKi/Gr/Vulkan/GrManagerImpl.cpp:1023]: I think this code should go inside GrManagerImpl::initDevice in the VRS section

File: AnKi/Renderer/VrsSriGeneration.cpp [hunk 3]
```diff
@@ -37,6 +37,7 @@ Error VrsSriGeneration::initInternal()
 		return Error::NONE;
 	}
 
+	m_sriTexelDimension = getGrManager().getDeviceCapabilities().m_minSriTexelSize <= 8 ? 8 : 16;
```
Comment (godlikepanos) [inline AnKi/Renderer/VrsSriGeneration.cpp:40]: according to the code style
```
Always place the **condition of a conditional ternary operator** inside parenthesis.

	// NO!!!!!!!!!!!
	a = b ? 1 : 0;

	// YES!!!!!!!!!!!!!!
	a = (b) ? 1 : 0;
```

**Review by godlikepanos — CHANGES_REQUESTED** [review]

File: AnKi/Gr/Vulkan/GrManagerImpl.cpp (hunk 1 above)
Comment (AlbinBernhardssonARM) [reply AnKi/Gr/Vulkan/GrManagerImpl.cpp:1031]: Going up to 32x32 is fine as long as 16x16 is still supported.
To check if 8x8 or 16x16 is supported is:
`(minFragmentShadingRateAttachmentTexelSize <= 16 <= maxFragmentShadingRateAttachmentTexelSize) || (minFragmentShadingRateAttachmentTexelSize <= 8 <= maxFragmentShadingRateAttachmentTexelSize)`
which can be simplified to:
`minFragmentShadingRateAttachmentTexelSize <= 16 && maxFragmentShadingRateAttachmentTexelSize >= 8`
Inverse is
`minFragmentShadingRateAttachmentTexelSize > 16 || maxFragmentShadingRateAttachmentTexelSize < 8`

Comment (godlikepanos) [reply AnKi/Gr/Vulkan/GrManagerImpl.cpp:1031]: nvm. I missed the "min" in minFragmentShadingRateAttachmentTexelSize

**Pull request merged**